### DISCUSSION TITLE
Foreign functions and Java 21

### DISCUSSION THREAD

**Discussion by ptziegler** [eclipse/windowbuilder#800] [General]
This is something that came to mind while recompiling the Windows binaries...

WindowBuilder makes heavy use of native methods, in order to take screenshots of the SWT widgets. Especially with the Linux binaries, there were (and probably still are) several issues where the IDE just crashes, especially on Wayland, with a segmentation fault.

I wonder if it would improve stability by switching from JNI to the Foreign Functions API. This would -however- raise the BREE for the OS bundles to Java 21.

Thoughts?

**Comment by merks** [comment]
All the EPP packages already require Java 21 for the 2024-06 release and so do linuxtools and wildwebdeveloper so it seems fine to go down this path.

**Reply by ptziegler** [reply]
So we wouldn't be the first ones to do so. That's already good to know. 😅

**Comment by ptziegler** [comment]
Nevermind... Foreign Functions are still only a [preview feature](https://docs.oracle.com/en/java/javase/21/core/foreign-function-and-memory-api.html) in Java 21 and were officially released with Java 22. So it'll probably take a few more years before this can be adopted.

**Reply by merks** [reply]
Yes, this discussion has come up in the Platform too where folks would like to use this to avoid maintaining native crappy things.
